File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum=0;
        for(int i:nums){
            sum+=i;
        }
        if(sum%2==1){
            return false;
        }
        int h=sum/2;
        return h(0,h,nums,0);
    }
    boolean h(int i,int h, int[] nums, int s){
        if(s==h){return true;}
        if(i>=nums.length){ return false;}
        return h(i+1,h,nums,s+nums[i]) || h(i+1,h,nums,s);
    }
}

File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans=[]
        def backtrack(start):
            if start==len(nums):
                ans.append(nums[:])
                return
            for i in range(start, len(nums)):
                nums[start],nums[i]=nums[i],nums[start]
                backtrack(start+1)
                nums[start],nums[i]=nums[i],nums[start]
        backtrack(0)
        return ans      
# from typing import List

# class Solution:
#     def permute(self, nums: List[int]) -> List[List[int]]:
#         ans = []
        
#         def create(temp, index_set):
#             if len(index_set) == 0:
#                 ans.append(temp[:])  # Add a copy of temp to ans
#                 return
            
#             for i in index_set:
#                 new_temp = temp + [nums[i]]  # Create a new temp list
#                 new_index_set = index_set - {i}  # Create a new index_set without i
#                 create(new_temp, new_index_set)
        
#         index_set = set(range(len(nums)))
#         create([], index_set)
#         return ans
